### PR TITLE
#26 Add the plyr library

### PR DESCRIPTION
**Pull request by rmdstudio**
- add plyr library assets
- add player template helper

### CODE REVIEW

File: template/helper/player.php [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+class ComFilesTemplateHelperPlayer extends KTemplateHelperAbstract
+{
+    protected static $HTML5_VIDEO_FORMATS = array('mp4', 'webm', 'ogg');
+
+    protected static $HTML5_AUDIO_FORMATS = array('aac', 'mp3', 'ogg', 'flac','x-flac', 'wave', 'wav', 'x-wav', 'x-pn-wav');
+
+    protected static $declared = false;
+
+    /*
+    public function load()
```
Comment (johanjanssens) [inline template/helper/player.php]: @rmdstudio Can you please clean up the code and remove this?

File: template/helper/player.php [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+/**
+ * @package    DOCman
+ * @copyright   Copyright (C) 2011 - 2014 Timble CVBA (http://www.timble.net)
+ * @license     GNU GPLv3 <http://www.gnu.org/licenses/gpl.html>
+ * @link        http://www.joomlatools.com
+ */
+
+class ComFilesTemplateHelperPlayer extends KTemplateHelperAbstract
+{
+    protected static $HTML5_VIDEO_FORMATS = array('mp4', 'webm', 'ogg');
```
Comment (johanjanssens) [inline template/helper/player.php]: @rmdstudio I would rename this to VIDEO_FORMATS, this is not related to HTML5 per se.

File: template/helper/player.php [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+    protected static $HTML5_AUDIO_FORMATS = array('aac', 'mp3', 'ogg', 'flac','x-flac', 'wave', 'wav', 'x-wav', 'x-pn-wav');
```
Comment (johanjanssens) [inline template/helper/player.php]: @rmdstudio I would rename this to AUDIO_FORMATS, this is not related to HTML5 per se.

File: template/helper/player.php [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    */
+
+    protected function _initialize(KObjectConfig $config)
+    {
+        $this->addDeclarations();
+
+        parent::_initialize($config);
+    }
+
+    public function addDeclarations()
```
Comment (johanjanssens) [inline template/helper/player.php]: @rmdstudio Why do you have moved this into a separate method which is being called during initialise? Does this need to be a public method of the helper?

File: template/helper/player.php (hunk 2 above)
Comment (rmdstudio) [inline template/helper/player.php]: Understood, I'll change that

File: template/helper/player.php (hunk 1 above)
Comment (rmdstudio) [inline template/helper/player.php]: The reason I kept that was because @ercanozkaya suggested that I insert that output in layouts. However that approach doesn't work when we need to use the template in the content plugin, so instead I created a singleton that imports the assets only once. I left the commented out code so Ercan would know about the change. I was going to clean it up afterwards.

File: template/helper/player.php (hunk 4 above)
Comment (rmdstudio) [inline template/helper/player.php]: I'm still undecided on this. Would you rather I moved the code block in the initialize and get rid of the public method? Because the initialize approach seems to be working fine. I wrote it as a separate function to keep the code organized. We can make it protected if you like.

**Comment by rmdstudio** [comment]
@johanjanssens  I applied the changes that you have requested and cleaned up the code

File: template/helper/player.php (hunk 1 above)
Comment (ercanozkaya) [inline template/helper/player.php]: @rmdstudio what's the issue with using the template in the content plugin? We still render templates there so you can use helper calls?

File: template/helper/player.php [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+    protected static $_DECLARED = false;
+
+    protected function _initialize(KObjectConfig $config)
+    {
+        if (! self::$_DECLARED) {
+
+            $document = JFactory::getDocument();
+            $document->addScript('media/koowa/com_files/js/plyr/plyr.js');
+            $document->addScriptDeclaration('jQuery(function($){plyr.setup()});');
+            $document->addStylesheet('media/koowa/com_files/css/plyr.css');
```
Comment (ercanozkaya) [inline template/helper/player.php]: @rmdstudio Why are we using Joomla API instead of `<ktml:script>` tags?

Comment (rmdstudio) [inline template/helper/player.php]: @ercanozkaya yes, because we want to be able to use this singleton in the content plugin as well and not just layouts.

File: template/helper/player.php (hunk 1 above)
Comment (rmdstudio) [inline template/helper/player.php]: @ercanozkaya in the FILElink content plugin I am replacing the urls with players and the plyr declarations is added to the header of the page only once. 

At first I ended up with 2 methods one using JDocument for the content plugin and another inserting <ktml:script> for the layouts. I had a complex logic to make sure the plyr declarations wouldn't happen twice. Then I decided to roll back to using a JDocument singleton, since I didn't know any KDocument way of inserting css and js declarations to the html header.

Comment (rmdstudio) [inline template/helper/player.php]: Also I'm using the term singleton correctly here. The idea is that the css and js get included only once.

File: template/helper/player.php (hunk 5 above)
Comment (ercanozkaya) [inline template/helper/player.php]: @rmdstudio Content plugin can still render templates we do not need to use the Joomla API. See Doclink plugin in Docman for an example.

Comment (rmdstudio) [inline template/helper/player.php]: @ercanozkaya DOClink doesn't say much. Question: if I put the js and css declarations in `views/plyr/tmp/default.php` will I be able to load it using `$this->getObject('com:files.view.plyr.html')->render();` or similar concept?

File: view/player/tmpl/default.html.php [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+ * @license		GNU GPLv3 <http://www.gnu.org/licenses/gpl.html>
+ * @link		http://github.com/nooku/nooku-files for the canonical source repository
+ */
+defined('KOOWA') or die;
+?>
+
+<ktml:style src="assets://files/css/plyr.css" />
+<ktml:script src="assets://files/js/plyr/plyr.js" />
+<script>
+    jQuery(function($){
```
Comment (ercanozkaya) [inline view/player/tmpl/default.html.php]: @rmdstudio Few things: 
* You should call `<?= helper('behavior.jquery') ?> on this file as it uses jQuery. 
* You should use `kQuery` instead of `jQuery` as the variable name. This makes sure that we only use on jQuery instance. (Joomla templates are known to include multiple)
* plyr.js should also use kQuery instead of jQuery. 

Let me know if you have any questions about the usage of kQuery. You can find examples on `KTemplateHelperBehavior`

File: template/helper/player.php [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+    {
+        if (! self::$_DECLARED) {
+
+            $template = $this->getObject('com:files.view.plyr.html')
+                            ->getTemplate()
+                            ->addFilter('style')
+                            ->addFilter('script')
+                            ->loadFile('com:files.player.default.html');
+
+            print $template->render();
```
Comment (ercanozkaya) [inline template/helper/player.php]: @rmdstudio Cool, I like the use of KTemplate. You shouldn't use print in any template helper. The code that outputs HTML should also not live in _initialize. Instead create a `display` method (just like editor.display method) that would return the HTML. 

The client code should use helper('player.display') or `$helper->display()` directly.

Comment (johanjanssens) [inline template/helper/player.php]: @ercanozkaya We did rename the whole rendering pipeline to use 'render' in the framework and not 'display' maybe we should use this in the helpers to as the default?

Comment (rmdstudio) [inline template/helper/player.php]: @ercanozkaya in the new commit I moved the import code into an import function. I also removed the print statement, because it was unnecessary. I am also using kQuery everywhere instead of jQuery.

Comment (ercanozkaya) [inline template/helper/player.php]: @johanjanssens Display is everywhere in JFW. Let's rename them all in one go later?

@rmdstudio can you still move the code into a public display function and not use _initialize to import it please? This makes sure the helper class is useful for more later on as needed.

Comment (johanjanssens) [inline template/helper/player.php]: @ercanozkaya Is it? I could only find one reference to display() or did we only change this in Kodekit?

Comment (rmdstudio) [inline template/helper/player.php]: @ercanozkaya The reason I didn't make it a display method is because it isn't displaying anything. It is importing the js and css in html header and initializing the UI. That's why it is part of the _initialize method. Also it is not an optional method. The UI needs to be initiated at least once and only once, that's why it is automatically called rather than us calling it every time the behaviour is used? That said, I'd still change it if that's what you want.

Comment (ercanozkaya) [inline template/helper/player.php]: @johanjanssens Sorry, I missed that we ported the changes into JFW. You are right. 

@rmdstudio Fair enough, in that case we can use the same paradigm as KTemplateHelperUi and call it `player.load` ?

Comment (johanjanssens) [inline template/helper/player.php]: Yep load sounds great, if we are already using that. Ideally it can be called include() but we cannot have that method, so lets go for load.

Comment (rmdstudio) [inline template/helper/player.php]: @ercanozkaya @johanjanssens I changed the protected method to public load. I also removed the initialize method and now explicitly calling player.load in the layouts and $helper->load in the plugin

**Comment by ercanozkaya** [comment]
@rmdstudio looking good to me.

**Comment by rmdstudio** [comment]
@ercanozkaya @johanjanssens are you going to approve the merge now?

**Comment by johanjanssens** [comment]
@rmdstudio See my last comments.

File: view/player/tmpl/default.html.php [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+ * @link		http://github.com/nooku/nooku-files for the canonical source repository
+ */
+defined('KOOWA') or die;
+?>
+
+<?= helper('behavior.koowa'); ?>
+
+<ktml:style src="assets://files/css/plyr.css" />
+<ktml:script src="assets://files/js/plyr/plyr.js" />
+<script>
```
Comment (rmdstudio) [inline view/player/tmpl/default.html.php]: @johanjanssens they were originally included using JDocument but per discussion we moved them into a layout. They need to be included to the page headers and so far the only ways that I know of are either JDocument or khtml tags in a layout. Is there a koowa way of including them in headers within the load function instead of importing a template layout?

File: view/player/tmpl/default.html.php [hunk 9]
```diff
@@ -0,0 +1,23 @@
+<?php
+/**
+ * Nooku Framework - http://nooku.org/framework
+ *
+ * @copyright	Copyright (C) 2011 - 2014 Johan Janssens and Timble CVBA. (http://www.timble.net)
```
Comment (rmdstudio) [inline view/player/tmpl/default.html.php]: Yes all the other files have the same header even in the master branch. What do I need to change it into?

Comment (johanjanssens) [inline view/player/tmpl/default.html.php]: It's oki. This is in com:files. Need to change all the headers here. @ercanozkaya We need to make tickets for this across all repo's and update headers. Remove Nooku references update dates etc.

File: view/player/tmpl/default.html.php (hunk 8 above)
Comment (johanjanssens) [inline view/player/tmpl/default.html.php]: Oki great :)
